library/core/src/pin.rs

```diff
@@ -1003,22 +1003,26 @@ impl<P, U> CoerceUnsized<Pin<U>> for Pin<P> where P: CoerceUnsized<U> {}
 #[stable(feature = "pin", since = "1.33.0")]
 impl<P, U> DispatchFromDyn<Pin<U>> for Pin<P> where P: DispatchFromDyn<U> {}
 
-/// Constructs a <code>[Pin]<[&mut] T></code>, by pinning[^1] a `value: T` _locally_[^2].
+/// Constructs a <code>[Pin]<[&mut] T></code>, by pinning a value locally.
 ///
-/// Unlike [`Box::pin`], this does not involve a heap allocation.
+/// Unlike [`Box::pin`], this does not involve a heap allocation. As explained
+/// below, the element might still end up on the heap however.
 ///
-/// [^1]: If the (type `T` of the) given value does not implement [`Unpin`], then this
-/// effectively pins the `value` in memory, where it will be unable to be moved.
-/// Otherwise, <code>[Pin]<[&mut] T></code> behaves like <code>[&mut] T</code>, and operations such
-/// as [`mem::replace()`][crate::mem::replace] will allow extracting that value, and therefore,
-/// moving it.
-/// See [the `Unpin` section of the `pin` module][self#unpin] for more info.
+/// The local pinning performed by this macro is usually dubbed "stack"-pinning.
+/// While local values are almost always located in the stack (e.g. when within
+/// the body of a non-`async` function), in `async` contexts or in the body of a
+/// generator this isn't neccessarily the case. Here, any locals crossing an
+/// `.await`/yield point are part of the state captured by the
+/// `Future`/Generator, and will use the storage of those, which typically means
```

Comment:
> which typically means heap storage

I feel this is a bit misleading. It means heap storage only if the future/generator where this macro is invoked is moved to the heap. This is admittedly often the case, but I fear the current phrasing might cause people to believe this macro performs boxing in and of itself.

Perhaps it could be modified to
> Which, if the future/generator is boxed, means heap storage

Reply:
Often times it's completely hidden from the user what is happening. You just await up until the top where you have `#[tokio::main]` or whatever. Internally it might use heap storage, but it's hidden from the user. So I'm not sure if "if the future/generator is boxed" is going to help much. I added the sentence to make it clear that non-stack options exist so "locally" doesn't automatically imply stack.

Reply:
I think we should actively avoid a reference to "stack-pinning" here if we can, as it has a "Don't think about the pink elephant" quality to it. Instead, it might be best to express that "local" is relative to whatever is actually being pinned, thus it uses the same storage (and implies the same lifetime).

Reply:
I think for programmers coming from high level languages it's useful to explain the distinction to them, because it's a great opportunity for them to learn about it. If we hide this detail from them then where else would they find out about it? It shouldn't be some secret that people "in the know" know. It might seem obvious to you and me, but just a few days ago I found these two paragraphs to contain very interesting information, at least after I managed to parse them :).

Reply:
I don't think that a "don't think of the pink elephant" situation applies here, because it's not wrong to *think* about it as stack storage. On a high level it looks like stack storage, as that's the point of `async` fn and blocks. It's only on a lower level where the stack analogy is wrong. And it's just as wrong for any other local that lives past an await boundary, it's just that, correct me if I'm wrong, one of the big use cases of `Pin` is such situations.

Reply:
I am speaking of myself when I discuss people who started as JavaScript programmers (a relatively new one to boot) and went into Rust and found the existing documentation on things like heap and stack strewn about the stdlib docs frustrating and confusing: I think it is familiarity bias that suggests that terms like the "stack" and "heap" are lightweight and easy to teach, such that it can be easily shimmed inside documents about something else. They're jargon.

What a Rust programmer is really concerned about (will be forced to care about!) is the lifetime. And as a JavaScript programmer, the idea that a "stack frame" existed and that it bounded the lifetime of something on the stack was not a simple thing to get across. I think I didn't learn that until after... a year of Rust?

So I am not trying to suggest hiding a detail. I just am saying that it was not a useful point of reference.

Reply:
@workingjubilee I see. the people you describe and used to belong to are one of the audiences that the docs serve and have to be accesible to. Other audiences exist as well, like people who know what stack and heap are and want to know how `pin!` relates to those two concepts: that's why I think the section is useful. I believe that the docs should be as jargon free as possible but one shouldn't sacrifice information content for it. I think stack and heap are pretty common things often taught in undergrad computer science classes and non-college programming courses. *technically* they might be jargon as the origin of the word isn't as connected with the concept itself as in other places in programming, but they are jargon of the most innocuous kind.

However, all this attention made me wonder now about the usually dubbed stack pinning part. I've thought it's an established term outside of rust. If it were, one could argue that it's important to mention the stack so that people who know stack pinning can easily reach this by google and also easily verify that it's the feature they are looking for. Looking up stack pinning doesn't give me any non-rust usage of such terms, outside of maybe [pin_ptr](https://learn.microsoft.com/en-us/cpp/extensions/pin-ptr-cpp-cli?view=msvc-170). @danielhenrymantilla as you added these docs in #93176, what does the "usually" mean, which places have been calling it "stack" pinning? Are there books/lectures/papers/other languages that call it that way? Maybe one should remove the first sentence?

Reply:
I guess https://docs.rs/futures/0.3.27/futures/macro.pin_mut.html and https://docs.rs/tokio/1.26.0/tokio/macro.pin.html may have been the main examples that gave me that impression

Reply:
I wasn't taught about stack and heap _despite_ taking programming courses. :/ I had to self-study that stuff. All of it.

( And the stack is a form of allocation anyways, just using hardware instructions... )

I don't mind using the words stack and heap per se, because I agree: for people who are familiar with that abstraction, they're a useful context. I just want to stress that docs should stand without them and address language-specific concerns first.